File: _utils/list.py
from typing import List


class ListNode:
    """Head of a singly-linked list.
    """
    def __init__(self, val: any, next = None) -> None:
        """Constructor of the singly-linked list head.
        Args:
            val: value of the node.
            next: next node in the list.
        """
        self.val = val
        self.next = next


class SinglyLinkedList:
    """The singly-linked list.
    """
    def __init__(self, head) -> None:
        self.head = head
        self.size = 0
        if head is not None:
            if isinstance(head, ListNode):
                self.head = head
            else:
                self.head = ListNode(head)
            self.size += 1
    def __str__(self):
        return SinglyLinkedList.printList(self.head)


    @staticmethod
    def array2list(vals: None) -> ListNode | None:
        """Create a list out of the given iterable.
        Args:
            vals: the iterable of values for the list.
        Returns:
            the head of the list.
        """
        if vals is None or len(vals) == 0:
            return None
        head = None
        cur = None
        for val in vals:
            if head is None:
                cur = ListNode(val)
                head = cur
            else:
                cur.next = ListNode(val)
                cur = cur.next
        return head


    @staticmethod
    def printList(head: ListNode | None, toString = None) -> str:
        """Print the singly-linked list.
        Args:
            head: first node of the list.
            toString: the custom function to print each node.
        Returns:
            the string representation of the list.
        """
        def _print(val) -> str:
            try:
                return toString(val)
            except:
                return str(val)

        if head is None:
            return _print(head)
        connector = " -> "
        re = ""
        while True:
            if head is not None:
                re += _print(head.val) + connector
                head = head.next
            elif re.endswith(connector):
                return re[:-len(connector)]
            else:
                return _print(head)


def main() -> None:
    import json

    vals = None
    head = SinglyLinkedList.array2list(vals)
    print(f"vals={json.dumps(vals)}")
    print(SinglyLinkedList.printList(head))

    vals = []
    head = SinglyLinkedList.array2list(vals)
    print(f"vals={json.dumps(vals)}")
    print(SinglyLinkedList.printList(head))

    vals = [1, 3, 5, 7, 9, 2, 4, 6, 8]
    head = SinglyLinkedList.array2list(vals)
    print(f"vals={json.dumps(vals)}")
    print(SinglyLinkedList.printList(head))

    print(SinglyLinkedList(head))


# if __name__ == "__main__":
#     main()

File: _utils/list.js
/**
 * Head of a singly-linked list
 * @param {object}    val     value of the node in the list
 * @param {ListNode}  next    next node in the list
 */
const ListNode = module.exports.ListNode = class {
    constructor(val, next) {
        this._val = val;
        this._next = next || null;
    }
    get val() { return this._val; }
    set val(v) { this._val = v; }
    get next() { return this._next; }
    set next(node) { this._next = node; }
};


/**
 * The singly-linked list
 * @param {object}  head    head (value) of the list.
 */
const SinglyLinkedList = module.exports.SinglyLinkedList = class {
    constructor(head) {
        this._size = 0;
        this._head = null;
        if (head !== null && head !== undefined) {
            this._head = head instanceof(ListNode) ? head : new ListNode(head);
            this._size = 1;
        }
    }
    get head() { return this._head; }
    get size() { return this._size; }
    toString() { return SinglyLinkedList.printList(this._head); }


    /**
     * create a list out of an array
     * @param {object[]}    vals      the number array
     * @returns {ListNode}            the head of the created list
     */
    static array2list(vals) {
        if (!vals || vals.length === 0) {
            return null;
        }
        const head = new ListNode(vals[0]);
        let cur = head;
        for (let i = 1; i < vals.length; i++) {
            cur = cur.next = new ListNode(vals[i]);
        }
        return head;
    }


    /**
     * Print the singly-linked list
     * @param {ListNode}  head      first node of the list
     * @param {function}  toString  the custom function to print each node
     * @returns {string}            the string representation of the list
     */
    static printList(head, toString) {
        const _print = (val) => {
            try {
                return toString(val);
            } catch (_) {
                return `${val}`;
            }
        };
        const connector = " -> ";
        let re = "";
        while (true) {
            if (head) {
                re += _print(head.val) + connector;
                head = head.next;
            } else if (re.endsWith(connector)) {
                return re.substring(0, re.length - connector.length);
            } else {
                return _print(head);
            }
        }
    };
};


function main() {
    let vals = null;
    let head = SinglyLinkedList.array2list(vals);
    console.log(`vals=${JSON.stringify(vals)}`);
    console.log(SinglyLinkedList.printList(head));

    vals = [];
    head = SinglyLinkedList.array2list(vals);
    console.log(`vals=${JSON.stringify(vals)}`);
    console.log(SinglyLinkedList.printList(head));

    vals = [1, 3, 5, 7, 9, 2, 4, 6, 8];
    head = SinglyLinkedList.array2list(vals);
    console.log(`vals=${JSON.stringify(vals)}`);
    console.log(SinglyLinkedList.printList(head));

    console.log(`${new SinglyLinkedList(head)}`);
}

main();
